### PR TITLE
OpenACC Model

### PR DESCRIPTION
**Pull request by pranav-sivaraman**
We have created an OpenACC implementation of Cloverleaf based on the OpenMP Offload Implementation

cc: @jhdavis8

### Pragma Conversions

| OpenMP | OpenACC |
| ------------- | ------------- |
| `omp target`  | `acc`  |
| `teams distribute parallel for simd`  | `parallel loop gang worker vector`  |
| `map(alloc : ...)` | `create(...)`
| `from(...)` | `host(...)` (this gets converted to self)
| `use_device_ptr(...)` | `deviceptr(...)`
| `map(release:...)` | `delete(...)`

For every accelerated loop there is also a `default(present)` clause to indicate that the data is on the GPU and not generate any extra copies.

### Function Conversions
| OpenMP | OpenACC |
| ------------- | ------------- |
| `omp_get_num_devices()`  | `acc_get_num_devices(acc_device_default)`  |
| `omp_set_default_device()` | `acc_set_device_num(..., acc_device_default)` |
| `omp_get_default_device()` | `acc_get_device_num(acc_device_default)` |

Tested with NVHPC@22.7 on V100 and A100 and got similar performance results to the OpenMP version.

Any feedback would be greatly appreciated

### CODE REVIEW

**Comment by tom91136** [comment]
Thanks for the PR! I'll review it shortly. Is there a RAJA version as well? I saw a RAJA commit from your fork which looks very interesting.

**Comment by pranav-sivaraman** [comment]
Yes we have a RAJA version as well!

Right now I am having issues just getting stuff to consistently build with the whole RAJA suite (RAJA/CHAI,Umpire) across all of our RAJA implementations. Especially if the RAJA suite is externally installed (with Spack). So we are waiting on getting feedback from the RAJA developers before we proceed with any PRs involving RAJA.

File: src/acc/calc_dt.cpp
Comment (tom91136) [inline src/acc/calc_dt.cpp]: You mean `reduction(min: dtct, dtut, dtvt, dtdivt, dt_min_va)` and then a final `fmin(dt_min_val, fmin(dtct, fmin(dtut, fmin(dtvt, fmin(dtdivt, g_big)))));` on the host? I would be interested to see the difference if you have time to benchmark this, preferably using `clover_bm16` or larger, ideally  `clover_bm64` or if memory allows `clover_bm256`.

File: src/acc/generate_chunk.cpp
Comment (tom91136) [inline src/acc/generate_chunk.cpp]: In this particular case I'm not too worried because it's only used a few times for initialisation so we're just ensuring the data resides at the right place.
Please elaborate on what you mean by forcing it.

File: src/acc/pack_kernel.cpp
Comment (tom91136) [inline src/acc/pack_kernel.cpp]: This one is slightly more concerning, does `collapse(2)` help? 
One of the OpenMP target implementations we've tested (it may have been vanilla Clang) had some performance issues with `collapse(2)`  in packing/unpacking so we didn't write it that way.

File: src/acc/comms_kernel.cpp
Comment (tom91136) [inline src/acc/comms_kernel.cpp]: We have a bug here if we 1) use a CUDA-aware MPI or 2) specify `--staging true`  in the CLI and have >1 MPI ranks.
On NVHPC 23.9 (CUDA 12.2) with the bundled HPC-X MPI, there's a ~35% deviation from the solution on `clover_bm16_short.in`.

I'm not too well versed in OpenACC, but to pass validation, I made the following changes:

1. Replace this directive with `#pragma acc enter data`, drop the extra `{}` (`acc exit data` stays the same).
2. Replace all uses of `#pragma acc data if (!stage) deviceptr(...)` with `#pragma acc host_data if (!stage) use_device(...)`, as we're doing an CUDA-aware MPI call in the block, not a device-side kernel.

Can you double check whether you can reproduce this?

**Review by tom91136 — CHANGES_REQUESTED** [review]
Thanks again for the PR, this is looking good.
Just a few minor TODOs and the bug in `comms_kernel`. 
Once resolved, please retest with >1 MPI ranks (can be on the same GPU, performance will be bad)  and try both `--staging true` and `--staging false`.

File: src/acc/comms_kernel.cpp
Comment (pranav-sivaraman) [reply src/acc/comms_kernel.cpp]: Are you sure you get a ~35% deviation when you specify `--staging-buffer true`? I get that ~35 deviation when I do a `--staging-buffer false`. I have attached my [output](https://github.com/UoB-HPC/CloverLeaf/files/13233350/output.txt).

Otherwise your changes did indeed pass validation for both `true` and `false` for `--staging-buffer` on `clover_bm16_short.in`.

File: src/acc/calc_dt.cpp
Comment (pranav-sivaraman) [reply src/acc/calc_dt.cpp]: Yes that's what I would like to try. Thanks for the suggestions for the input sizes. I will definitely contact you with any findings I have.

File: src/acc/pack_kernel.cpp
Comment (pranav-sivaraman) [reply src/acc/pack_kernel.cpp]: I am not exactly sure what the impact of `collapse(2)` will have on performance. I would probably need to do profiling to fully investigate. It does pass validation with `clover_bm16_short.in` if I do add a `collapse(2)`. 

We could add the `collapse(2)` just for OpenACC since NVHPC seems to be the only compiler with a mature implementation. 

As for the OpenMP version maybe there are performance issues with vanilla Clang due to `simd` not being fully implemented in `teams distribute parallel for simd`? I think we should add the `collapse(2)` to the OpenMP version if I am assuming the NVHPC compiler does not have performance issues. Hopefully vanilla clang's OpenMP implementation progresses where this won't become an issue in the future?

File: src/acc/generate_chunk.cpp
Comment (pranav-sivaraman) [inline src/acc/generate_chunk.cpp:134]:
```suggestion
              #pragma acc loop collapse(2)
              for (int kt = j; kt <= j + 1; ++kt) {
```
Since NVHPC could not parallelize the following loops from its own analysis we can "force" the compiler to parallelize the loops by adding a  `#pragma acc loop collapse(2)` to all the nested for loops. 

I passed validation on `clover_bm16_short.in`, but I am not sure if other input decks may result in an error.

**Comment by pranav-sivaraman** [comment]
I've implemented the bug fixes regarding `comms_kernel` and have verified that it's passing the chosen test.

I would like add an option for using managed memory for OpenACC, but NVHPC also supports it for OpenMP as well. Should this be a separate PR? If it is would it make sense to remove the CPU options for OpenACC? TBH I am not too well versed on the CPU options so maybe instead just refactor to make it like OMP Offload? That way we can also add flags for `gcc` as well for both AMD and NVIDIA gpus.

File: src/acc/comms_kernel.cpp
Comment (tom91136) [reply src/acc/comms_kernel.cpp]: Sorry, it was a typo, `--staging-buffer true` forces device-to-host staging so it's the safe path.

**Comment by tom91136** [comment]
Yes, please open a separate PR for changes outside of OpenACC. I think the current OpenACC implementation works fine on the host, but feel free to make any changes you think are appropriate and I'll review it.

File: src/acc/generate_chunk.cpp
Comment (tom91136) [reply src/acc/generate_chunk.cpp:134]: Be aware that only the following decks have checksums:
```
clover_bm.in
clover_bm16.in
clover_bm16_short.in
clover_bm16_very_short.in
clover_bm2_short.in
clover_bm4_short.in
clover_bm_short.in 
```
In general, if `clover_bm16.in` (this one has 3000 steps so is good at catching memory leaks and MPI related issues) is good at 1 and > 1 MPI ranks with and without staging, then the implementation is correct.

File: src/acc/pack_kernel.cpp
Comment (tom91136) [reply src/acc/pack_kernel.cpp]: Yes, please profile and keep `collapse(2)` if it helps. Open a separate PR if you think OpenMP will also benefit from this and the compiler doesn't fail or given wrong results.

**Comment by pranav-sivaraman** [comment]
@tom91136 I've added a managed memory option for OpenACC. As of right now I think its fine to merge in. Any other improvements would probably come with seperate PR.

One Note:
I've tested this with `gcc` OpenACC offloading and there appears to be a bug with the `default(present)` . I think the fix is replace all the `default(present)` with explicit present clauses. Not sure if it's worth doing since the better performance ends up lying in `NVHPC` anyways.

**Comment by pranav-sivaraman** [comment]
I've made some change to get the OpenACC model to work with GCC. I think this is ready to merge, but let me know if there are any other changes that need to be made.

**Comment by pranav-sivaraman** [comment]
Hmm after doing some testing on the `clover_bm16.in` input deck I noticed some changes causing NVHPC to fail. I have since reverted those and it is now passing, but GCC seems to be getting a runtime error in `calc_dt.cpp`

I get the following warning when compiling with GCC using the following flags `-fopenacc -foffload=nvptx-none='-march=sm_80 -foffload=-lm -foffload=-O3`
```
[1/2] Building CXX object CMakeFiles/cloverleaf.dir/src/acc/calc_dt.cpp.o
../src/acc/calc_dt.cpp: In function 'calc_dt_kernel(bool, int, int, int, int, double, double, double, double, double, field_type&, double&, int&, double&, double&, int&, int&, int&) [clone ._omp_fn.0]':
../src/acc/calc_dt.cpp:60:9: warning: 'dt_min_val.32' is used uninitialized [-Wuninitialized]
   60 | #pragma acc parallel loop gang worker vector collapse(2) clover_use_target(use_target) \
      |         ^~~
../src/acc/calc_dt.cpp:60:9: note: 'dt_min_val.32' was declared here
   60 | #pragma acc parallel loop gang worker vector collapse(2) clover_use_target(use_target) \
      |         ^~~
../src/acc/calc_dt.cpp:60:9: warning: 'dt_min_val.6' is used uninitialized [-Wuninitialized]
   60 | #pragma acc parallel loop gang worker vector collapse(2) clover_use_target(use_target) \
      |         ^~~
../src/acc/calc_dt.cpp:60:9: note: 'dt_min_val.6' was declared here
   60 | #pragma acc parallel loop gang worker vector collapse(2) clover_use_target(use_target) \
``` 

The runtime error is `CUDA Exception: Warp Illegal Address`. 

I'm not exactly sure what is the exact issue. Could it be related to some alignment issues? 

Edit: I've now fixed the issue by doing something similar to what was done in OpenMP Offload

**Comment by tom91136** [comment]
@pranav-sivaraman Sorry for the delay! I think this is ready for merge. Just to check, without  `present` , GCC still fails right? From what I understand, this works on GCC, and NVHPC currently.

**Comment by pranav-sivaraman** [comment]
With `default(present)`, `gcc@13` and before will fail. Adding the explicit `present` will cause it not to throw a runtime error. `gcc@14` says in their release notes they have now implemented `default(present)` for OpenMP/OpenACC, but I have not tested this.

**Comment by tom91136** [comment]
Thank you very much for this contribution @pranav-sivaraman and for putting up with this long delay. 
This brings us one step closer to the coverage BabelStream has.

**Pull request merged**
